src/shared/plugins/cookie-message/index.js

```diff
@@ -1,42 +1,162 @@
+ */
+const getCurrentPath = request => isEmpty(request.query)
+  ? request.path
+  : `${request.path}?${qs.stringify(request.query)}`;
 
-  return h.continue;
+/**
+ * Gets the path to set the cookie preferences
+ *
+ * @param {Object} request
+ * @param {Boolean} isAccepted - whether the user opts in/out of analytics cookies
+ * @returns {String}
+ */
+const getPreferencesPath = (request, isAccepted) => {
+  const redirectPath = getCurrentPath(request);
+  const query = qs.stringify({
+    redirectPath,
+    acceptAnalytics: isAccepted
+  });
+  return `/set-cookie-preferences?${query}`;
 };
 
+/**
+ * Gets the path to the cookies page, including a redirectPath query param which points
+ * back to the user's current path
+ *
+ * @param {Object} request
+ * @returns {String}
+ */
+const getCookiesPagePath = request => `/cookies?${qs.stringify({ redirectPath: getCurrentPath(request) })}`;
+
+/**
+ * Pre handler sets cookie banner state in view
+ *
+ * @param {Object} request
+ * @param {Object} h
+ */
 const _handler = async (request, h) => {
-  const { seen_cookie_message: seenCookieMessage } = request.state;
-  return (seenCookieMessage === 'yes') ? h.continue : setCookie(request, h);
+  const isEnabled = request.isAnalyticsCookiesEnabled();
+
+  // Get flash messages
+  const [flashMessage] = request.yar.flash(constants.flashMessageType);
+
+  set(request, 'view.cookieBanner', {
+    flashMessage,
+    isAnalyticsCookiesEnabled: isEnabled,
+    isVisible: (isEnabled === null) && !isCookiesPage(request),
+    acceptPath: getPreferencesPath(request, 1),
+    rejectPath: getPreferencesPath(request, 0),
+    cookiesPagePath: getCookiesPagePath(request)
+  });
+
+  return h.continue;
 };
 
+/**
+ * Checks whether analytics cookies are enabled by inspecting the state
+ * of the cookie.  If null is returned, the user has not yet
+ * set their preferences.
+ *
+ * @return {Boolean|Null}
+ */
+function isAnalyticsCookiesEnabled () {
+  const value = this.state[constants.cookieName];
+  if (value === constants.accepted) {
+    return true;
+  }
+  return value === constants.rejected ? false : null;
+}
+
+/**
+ * Sets the users analytics cookie preferences by setting the state
+ * of the cookie.
+ *
+ * If the user has opted out, we unset any Google Analytics cookies
+ * that are currently set.
+ *
+ * @param {Boolean} isAnalyticsAccepted
+ */
+function setCookiePreferences (isAnalyticsAccepted) {
+  // Set preferences
+  this.state(constants.cookieName, isAnalyticsAccepted ? constants.accepted : constants.rejected);
+
+  // Clear analytics cookies
+  if (!isAnalyticsAccepted) {
+    ['_ga', '_gid', '_gat', '_gat_govuk_shared'].forEach(cookieName => {
+      this.unstate(cookieName);
+    });
+  }
+}
+
+/**
+ * Gets the options for declaring the cookie that will manage the
+ * preferences
+ *
+ * @returns {Object}
+ */
+const getCookieOptions = () => ({
+  isSecure: process.env.NODE_ENV !== 'local',
+  isHttpOnly: true,
+  ttl: 365 * 24 * 60 * 60 * 1000,
```

Comment:
Should this be 28 days rather than 365?

Reply:
I looked at some other GOV.uk sites and they seemed to set it for a year, otherwise the banner will pop up monthly

Reply:
Cool. If that's the done thing then let's leave it 👍🏽